tools/walletextension/readwriter/read_writer.go

```diff
@@ -0,0 +1,102 @@
+package readwriter
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/obscuronet/go-obscuro/go/common/log"
+
+	"github.com/gorilla/websocket"
+)
+
+const (
+	headerUpgrade   = "Upgrade"
+	headerWebsocket = "websocket"
+	httpCodeErr     = 500
+)
+
+var upgrader = websocket.Upgrader{} // Used to upgrade connections to websocket connections.
+
+// ReadWriter handles reading and writing Ethereum JSON RPC requests.
+type ReadWriter interface {
+	ReadRequest() ([]byte, error)
+	WriteResponse(msg []byte) error
+	HandleError(msg string)
+	SupportsSubscriptions() bool
+}
+
+// HTTPReadWriter is a ReadWriter over HTTP.
+type HTTPReadWriter struct {
+	resp http.ResponseWriter
+	req  *http.Request
+}
+
+// WSReadWriter is a ReadWriter over websockets.
+type WSReadWriter struct {
+	conn *websocket.Conn
+}
+
+func NewReadWriter(resp http.ResponseWriter, req *http.Request) (ReadWriter, error) {
+	for _, header := range req.Header[headerUpgrade] {
```

Comment:
Would it be useful to add some comments on the header upgrade operations?

Reply:
Yep, added.